File: plugin/graph/src/main/groovy/net/twisterrob/gradle/graph/graphstream/GraphStreamTaskVisualizer.java
package net.twisterrob.gradle.graph.graphstream;

import java.awt.Window;
import java.awt.event.*;
import java.io.IOException;
import java.util.*;

import javax.swing.SwingUtilities;

import org.codehaus.groovy.runtime.IOGroovyMethods;
import org.gradle.api.*;
import org.gradle.cache.PersistentCache;
import org.graphstream.graph.*;
import org.graphstream.graph.implementations.SingleGraph;
import org.graphstream.ui.layout.Layout;
import org.graphstream.ui.layout.springbox.implementations.LinLog;
import org.graphstream.ui.swingViewer.ViewPanel;
import org.graphstream.ui.view.Viewer;

import net.twisterrob.gradle.graph.*;
import net.twisterrob.gradle.graph.graphstream.GraphStreamSettings.WindowLocation;

import static net.twisterrob.gradle.graph.graphstream.GraphExtensions.*;

public class GraphStreamTaskVisualizer implements TaskVisualizer {
	private Graph graph;
	private Viewer viewer;
	private final GraphStreamSettings settings;

	public GraphStreamTaskVisualizer(PersistentCache cache) {
		settings = new GraphStreamSettings(cache);
	}

	@Override public void showUI(Project project) {
		graph = new SingleGraph(project.getName());
		try {
			String css = IOGroovyMethods.getText(getClass().getResourceAsStream("/graphstream.css"));
			graph.addAttribute("ui.stylesheet", css);
		} catch (IOException e) {
			throw new IllegalStateException("Cannot read style sheet");
		}
		graph.setStrict(true);
		graph.setAutoCreate(false);

		Layout layout = new LinLog(false);
		viewer = graph.display();
		Window window = SwingUtilities.getWindowAncestor(viewer.getDefaultView());
		window.createBufferStrategy(1);
		viewer.setCloseFramePolicy(Viewer.CloseFramePolicy.CLOSE_VIEWER);
		window.addWindowListener(new WindowAdapter() {
			@Override public void windowClosing(WindowEvent e) {
				settings.setSettings(new WindowLocation(e.getWindow()));
			}
		});
		settings.getSettings().applyTo(viewer);
		//viewer.enableAutoLayout(layout);
	}

	private void createNode(TaskData data) {
		Node node = graph.addNode(id(data.getTask()));
		setLabel(node, data.getTask().getName());
		addClass(node, classMappingType.get(data.getType()));
	}

	@Override public void initModel(Map<Task, TaskData> tasks) {
		for (TaskData data : tasks.values()) {
			createNode(data);
		}
		for (TaskData data : tasks.values()) {
			Node from = graph.getNode(id(data.getTask()));
			for (TaskData dep : data.getDepsDirect()) {
				Node to = graph.getNode(id(dep.getTask()));
				graph.addEdge(id(data.getTask(), dep.getTask()), from, to, true);
			}
		}
	}

	@Override public void closeUI() {
		ViewPanel view = viewer.getDefaultView();
		if(view != null) {
			settings.setSettings(new WindowLocation(SwingUtilities.getWindowAncestor(view)));
			viewer.removeView(view.getId());
			viewer.close();
		}
	}

	@Override public void update(Task task, TaskResult result) {
		Node node = graph.getNode(id(task));
		for (Map.Entry<TaskResult, String> possibleResult : classMappingResult.entrySet()) {
			removeClass(node, possibleResult.getValue());
		}
		String[] classes = addClass(node, classMappingResult.get(result));
		//System.out.println(task.getName() + ": " + Arrays.toString(classes));
	}

	private static final EnumMap<TaskResult, String> classMappingResult = new EnumMap<>(TaskResult.class);
	private static final EnumMap<TaskType, String> classMappingType = new EnumMap<>(TaskType.class);

	static {
		classMappingResult.put(TaskResult.executing, "executing");
		classMappingResult.put(TaskResult.completed, "executed");
		classMappingResult.put(TaskResult.nowork, "nowork");
		classMappingResult.put(TaskResult.skipped, "skipped");
		classMappingResult.put(TaskResult.uptodate, "uptodate");
		classMappingResult.put(TaskResult.failure, "failure");
		classMappingType.put(TaskType.unknown, "unknown");
		classMappingType.put(TaskType.normal, "norma");
		classMappingType.put(TaskType.requested, "requested");
		classMappingType.put(TaskType.excluded, "excluded");
	}

	private static String id(Task task) {
		return task.getName();
	}

	private static String id(Task from, Task to) {
		return id(from) + "->" + id(to);
	}
}

File: graph/src/main/kotlin/net/twisterrob/gradle/graph/vis/graphstream/GraphStreamTaskVisualizer.kt
package net.twisterrob.gradle.graph.vis.graphstream

import net.twisterrob.gradle.graph.tasks.TaskData
import net.twisterrob.gradle.graph.tasks.TaskResult
import net.twisterrob.gradle.graph.tasks.TaskType
import net.twisterrob.gradle.graph.vis.TaskVisualizer
import org.gradle.api.Task
import org.gradle.api.initialization.Settings
import org.gradle.cache.PersistentCache
import org.graphstream.graph.Edge
import org.graphstream.graph.Graph
import org.graphstream.graph.Node
import org.graphstream.graph.implementations.MultiGraph
import org.graphstream.ui.view.Viewer
import java.awt.Component
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.io.IOException
import java.util.EnumMap
import javax.swing.SwingUtilities

class GraphStreamTaskVisualizer(cache: PersistentCache) : TaskVisualizer {

	@Suppress("LateinitUsage") // TODO use some kind of factory to make invalid state non-representable.
	private lateinit var graph: Graph

	@Suppress("LateinitUsage") // TODO use some kind of factory to make invalid state non-representable.
	private lateinit var viewer: Viewer

	private val options: Options = Options(cache)

	override fun showUI(settings: Settings) {
		System.setProperty("org.graphstream.ui", "swing") // 2.0
		graph = MultiGraph(settings.rootProject.name)
		try {
			val css = this::class.java.getResourceAsStream("/graphstream.css").bufferedReader().readText()
			graph.setAttribute("ui.stylesheet", css)
		} catch (ex: IOException) {
			throw IllegalStateException("Cannot read style sheet.", ex)
		}
		graph.isStrict = true
		graph.setAutoCreate(false)

		viewer = graph.display()
		val window = SwingUtilities.getWindowAncestor(viewer.defaultView as Component)
		window.createBufferStrategy(1)
		viewer.closeFramePolicy = Viewer.CloseFramePolicy.CLOSE_VIEWER
		window.addWindowListener(object : WindowAdapter() {
			override fun windowClosing(e: WindowEvent) {
				options.options = Options.WindowLocation(e.window)
				options.close()
			}
		})
		options.options.applyTo(viewer)
	}

	override fun initModel(graph: Map<Task, TaskData>) {
		viewer.disableAutoLayout()
		this.graph.setAttribute("layout.force", 1.0)
		for (data in graph.values) {
			val node = this.graph.addNode<Node>(id(data.task))
			node.label = data.task.path
			node.addClass(classMappingType.getValue(data.type))
			//node.setAttribute("layout.weight", @Suppress("MagicNumber") 100)
		}
		for (data in graph.values) {
			val from = this.graph.getNode<Node>(id(data.task))
			for (dep in data.depsDirect) {
				val to = this.graph.getNode<Node>(id(dep.task))
				val edge = this.graph.addEdge<Edge>(id(data.task, dep.task), from, to, true)
				edge.setAttribute("layout.weight", @Suppress("MagicNumber") 10)
			}
		}
		// See org.graphstream.ui.layout.springbox.BarnesHutLayout for what attributes it uses.
		viewer.enableAutoLayout()
	}

	override fun closeUI() {
		val view = viewer.defaultView
		if (view != null) {
			SwingUtilities.invokeLater {
				options.options = Options.WindowLocation(SwingUtilities.getWindowAncestor(view as Component))
				options.close()
				viewer.removeView(view.id)
				viewer.close()
			}
		}
	}

	override fun update(task: Task, result: TaskResult) {
		val node = graph.getNode<Node>(id(task))
		for (value in classMappingResult.values) {
			node.removeClass(value)
		}
		node.addClass(classMappingResult.getValue(result))
		//println(task.name + ": " + node.classes)
	}

	companion object {

		/** @see graphstream.css */
		private val classMappingResult: Map<TaskResult, String> =
			EnumMap<TaskResult, String>(TaskResult::class.java).apply {
				this[TaskResult.Executing] = "executing"
				this[TaskResult.Completed] = "executed"
				this[TaskResult.NoWork] = "nowork"
				this[TaskResult.Skipped] = "skipped"
				this[TaskResult.UpToDate] = "uptodate"
				this[TaskResult.Failure] = "failure"
				check(this.keys.size == TaskResult.values().size)
			}

		/** @see graphstream.css */
		private val classMappingType: Map<TaskType, String> =
			EnumMap<TaskType, String>(TaskType::class.java).apply {
				this[TaskType.Unknown] = "unknown"
				this[TaskType.Normal] = "normal"
				this[TaskType.Requested] = "requested"
				this[TaskType.Excluded] = "excluded"
				check(this.keys.size == TaskType.values().size)
			}

		@Suppress("FunctionMinLength")
		private fun id(task: Task): String =
			task.path

		@Suppress("FunctionMinLength")
		private fun id(from: Task, to: Task): String =
			id(from) + "->" + id(to)
	}
}
